feat(bulk-cdk): Add Dokka documentation generation with Vercel deployment

aaronsteers opened the pull request:
## What

Loom: https://www.loom.com/share/3daa6ab80fe74fa4ae97e3b4ee82789b

This PR implements automated API documentation generation for the Airbyte Bulk CDK using Dokka, with automatic deployment to Vercel for preview and production environments.

**Requested by:** AJ Steers (@aaronsteers)
**Devin session:** https://app.devin.ai/sessions/f46b19703bd848dbad2c58cb105d470a

This follows the implementation plan from a prior Devin session, adapting the PyAirbyte documentation pattern to the Kotlin-based Bulk CDK.

## How

### Gradle Configuration (`airbyte-cdk/bulk/build.gradle`)
- Added Dokka plugin v2.0.0 to the build
- Configured Dokka for all subprojects with:
  - Public API documentation only (excludes internal/private APIs)
  - External documentation links to Kotlin stdlib, kotlinx-coroutines, and Micronaut
  - Source links back to GitHub for easy navigation
  - Multi-module aggregation via `dokkaHtmlMultiModule` task
- Added convenience task `dokkaGenerate` for local development

### GitHub Actions Workflow (`.github/workflows/bulk-cdk-dokka-publish.yml`)
- **detect-changes**: Uses path filtering to only run when Bulk CDK files change
- **build-docs**: Generates Dokka documentation and uploads as artifact (runs for all PRs including forks)
- **vercel-preview**: Deploys to Vercel (skips for fork PRs to avoid secrets access issues)
  - Preview deployments for main repo PRs
  - Production deployment to `bulk-cdk-docs.vercel.app` on master branch pushes
  - Posts GitHub check with preview URL for easy access

## Review guide

### Critical items requiring setup before merge:
1. **Vercel Project Setup** (BLOCKING):
   - Create new Vercel project named "airbyte-bulk-cdk-docs"
   - Add `VERCEL_BULK_CDK_DOCS_PROJECT_ID` as a GitHub repository variable
   - Verify the domain `bulk-cdk-docs.vercel.app` is correct or update in workflow

2. **Gradle Configuration** (`airbyte-cdk/bulk/build.gradle`):
   - Lines 12-14: Dokka plugin declaration
   - Lines 87-122: Dokka configuration for subprojects
   - Lines 125-139: Multi-module task configuration
   - ⚠️ Note: Dokka v2.0.0 shows deprecation warning about V1 plugin mode - may need migration to V2 in future

3. **GitHub Actions Workflow** (`.github/workflows/bulk-cdk-dokka-publish.yml`):
   - Lines 20-33: Change detection logic
   - Lines 76-88: Fork PR protection (prevents Vercel deployment for forks)
   - Lines 114-125: Vercel deployment configuration
   - Lines 137-147: GitHub check posting with preview URL

### Testing checklist:
- [x] Local Dokka build succeeds (`./gradlew :airbyte-cdk:bulk:dokkaHtmlMultiModule`)
- [ ] Vercel project created and variable added
- [ ] Workflow runs successfully on this PR
- [ ] Preview deployment accessible
- [ ] Fork PR behavior (should skip Vercel deployment gracefully)
- [ ] Production deployment on master merge

## User Impact

**Positive:**
- Bulk CDK developers will have automatically generated, searchable API documentation
- Preview deployments on PRs allow reviewing documentation changes before merge
- External documentation links provide easy navigation to dependency docs

**Negative/Limitations:**
- Dokka plugin shows deprecation warning (V1 → V2 migration needed in future)
- Requires Vercel project setup before workflow will succeed
- Documentation only covers public APIs (internal/private APIs excluded by design)

## Can this PR be safely reverted and rolled back?

- [x] YES 💚


This PR only adds new functionality (documentation generation) and doesn't modify any existing Bulk CDK code or behavior. Reverting would simply remove the documentation workflow and Gradle configuration, with no impact on the CDK itself.

---

**Note:** This PR is marked "do not merge" until the Vercel project is created and the `VERCEL_BULK_CDK_DOCS_PROJECT_ID` variable is added to GitHub. The workflow will fail without this setup.

devin-ai-integration[bot] committed:
feat(bulk-cdk): Add Dokka documentation generation with Vercel deployment

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
docs(bulk-cdk): Add CONTRIBUTING.md with Dokka documentation instructions

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
refactor(bulk-cdk): Simplify CONTRIBUTING.md and rename task to docsGenerate

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
refactor(bulk-cdk): Rename to 'Kotlin Bulk CDK' in workflow, docs, and build config

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
style(bulk-cdk): Fix formatting in workflow file

Co-Authored-By: AJ Steers <aj@airbyte.io>

ian-at-airbyte commented on airbyte-cdk/bulk/CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+## Prerequisites
+
+- **Java 21** or higher
```
It took me a minute to figure out I needed a JDK v21 and not just Java itself. For a Java developer we can probably take that for granted, but I figure it's always good to be explicit (it tripped me up).

ian-at-airbyte reviewed (approved):
I was able to generate the docs and everything looked as expected. We should maybe look at building the output straight into Docusaurus as Markdown files, but that can be a separate thing. This LGTM (assuming we can repair the CI failures).

devin-ai-integration[bot] committed:
docs(bulk-cdk): Clarify JDK 21 requirement in CONTRIBUTING.md

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
style(bulk-cdk): Remove trailing whitespace from build.gradle

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
perf(bulk-cdk): Optimize workflow - enable build cache and reuse artifacts

- Remove --no-build-cache flag to enable Gradle build caching (saves 1-2 min)
- Have vercel-preview download artifact from build-docs instead of rebuilding (saves 5-6 min)
- Total expected time savings: ~6-8 minutes per workflow run

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
fix(bulk-cdk): Skip Vercel deployment when project ID not configured

Add conditional guard to skip vercel-preview job when VERCEL_BULK_CDK_DOCS_PROJECT_ID
variable is not set. This prevents CI failures while waiting for Vercel project setup.

The job will automatically deploy once the GitHub variable is configured.

Co-Authored-By: AJ Steers <aj@airbyte.io>

aaronsteers committed:
Merge branch 'master' into devin/1763583323-bulk-cdk-dokka-docs

aaronsteers commented on airbyte-cdk/bulk/CONTRIBUTING.md:
```diff
@@ -0,0 +1,46 @@
+
+- **JDK 21** (Java Development Kit) or higher
+- **Gradle** (uses the wrapper, no separate installation needed)
+
```
Devin, please add an L3 subsection "If you need to install Java"

The process we can recommend with `sdkman` would be something very simple:

```
# Get sdkman (https://sdkman.io/)
curl -s "https://get.sdkman.io" | bash
source "$HOME/.sdkman/bin/sdkman-init.sh"

# Verify install
sdk version

# Show available versions
sdk list java | grep 21

# Install the latest and set as default
sdk install java 21.0.9-zulu
sdk default java 21.0.9-zulu
```

devin-ai-integration[bot] committed:
docs(bulk-cdk): Add Java installation instructions using sdkman to CONTRIBUTING.md

Co-Authored-By: AJ Steers <aj@airbyte.io>

aaronsteers commented on .github/workflows/kotlin-bulk-cdk-dokka-publish.yml:
```diff
@@ -0,0 +1,137 @@
+        github.event_name == 'push'
+        || github.event.pull_request.head.repo.full_name == github.repository
+      )
+      && vars.VERCEL_BULK_CDK_DOCS_PROJECT_ID != ''
```
Aside - 

```suggestion
      && vars.VERCEL_KOTLIN_CDK_PROJECT_ID != ''
```

aaronsteers committed:
Apply suggestion from @aaronsteers

aaronsteers committed:
Apply suggestion from @aaronsteers

devin-ai-integration[bot] committed:
debug: Add artifact structure logging to diagnose Vercel path issue

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
fix(bulk-cdk): Download artifact to path matching Vercel Root Directory setting

The Vercel project has Root Directory set to 'airbyte-cdk/bulk', so the
artifact needs to be downloaded to docs-output/airbyte-cdk/bulk to match
that expectation. This fixes the deployment error:
'The provided path ~/work/airbyte/airbyte/docs-output/airbyte-cdk/bulk does not exist'

Also added verification step to confirm index.html is at the expected path.

Co-Authored-By: AJ Steers <aj@airbyte.io>

aaronsteers commented:
Devin, this PR is working well in the PR / preview context but fails on merge to main:

https://github.com/airbytehq/airbyte/actions/runs/19578468640

Please create a follow-on PR to fix the issue when run on master branch.
